agent/engine/engine_unix_integ_test.go

```diff
@@ -1533,24 +1542,94 @@ func TestFluentdTag(t *testing.T) {
+
+	A := createTestContainerWithImageAndName(baseImageForOS, "A")
+
+	A.EntryPoint = &entryPointForOS
+	A.Command = []string{"sleep 30"}
+	A.Essential = true
+
+	testTask.Containers = []*apicontainer.Container{
+		A,
+	}
+	execConfig := types.ExecConfig{
+		Detach: true,
+		Cmd:    []string{"ls"},
+	}
+	go taskEngine.AddTask(testTask)
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+
+	finished := make(chan interface{})
+	go func() {
+		// Both containers should start
+		verifyContainerRunningStateChange(t, taskEngine)
+		verifyTaskIsRunning(stateChangeEvents, testTask)
+
+		containerMap, _ := taskEngine.(*DockerTaskEngine).state.ContainerMapByArn(testTask.Arn)
+		dockerID := containerMap[testTask.Containers[0].Name].DockerID
+
+		//Create Exec object on the host
+		execContainerOut, err := taskEngine.(*DockerTaskEngine).client.CreateContainerExec(ctx, dockerID, execConfig, dockerclient.ContainerExecCreateTimeout)
+		require.NoError(t, err)
+		require.NotNil(t, execContainerOut)
+
+		//Start the above Exec process on the host
+		err1 := taskEngine.(*DockerTaskEngine).client.StartContainerExec(ctx, execContainerOut.ID, dockerclient.ContainerExecStartTimeout)
+		require.NoError(t, err1)
+
+		//Inspect the above Exec process on the host to check if the exit code is 0 which indicates successful run of the command
+		execContInspectOut, err := taskEngine.(*DockerTaskEngine).client.InspectContainerExec(ctx, execContainerOut.ID, dockerclient.ContainerExecInspectTimeout)
+		require.NoError(t, err)
+		require.Equal(t, dockerID, execContInspectOut.ContainerID)
+		require.Equal(t, 0, execContInspectOut.ExitCode)
+
+		// Task should stop
+		verifyContainerStoppedStateChange(t, taskEngine)
+		verifyTaskIsStopped(stateChangeEvents, testTask)
+		close(finished)
+	}()
+
+	waitFinished(t, finished, testTimeout)
 }
 
-// TODO: [ecs-exec] Enable this test when exec command agent is in place on the host
+// TestExecCommandAgent validates ExecCommandAgent start and monitor processes. The algorithm to test is as follows:
+// 1. Pre-setup: the make file in ../../misc/exec-command-agent-test will create a special docker sleeper image
+// based on a scratch image. This image simulates a customer image and contains pre-baked /sleep and /kill binaries.
+// /sleep is the main process used to launch the test container; /kill is an application that kills a process running in
+// the container given a PID.
+// The make file will also create a fake amazon-ssm-agent which is a go program that only sleeps for a certain time specified.
+//
+// 2. Setup: Create a new docker task engine with a modified path pointing to our fake amazon-ssm-agent binary
+// 3. Create and start our test task using our test image
+// 4. Wait for the task to start and verify that the expected ExecCommandAgent bind mounts are present in the containers
+// 5. Verify that our fake amazon-ssm-agent was started inside the container using docker top, and retrieve its PID
```

Comment:
why use docker top? can we do the same as in the code? i.e inspect docker exec ID?

Reply:
It would work, although that would imply that the integ test knows too much about the implementation, kinda more like a unit test. Docker top seems more high level and independent of how we execute the agent, al we care about is that it's running in our test container.